### PR TITLE
Make `--dev` the default

### PR DESCRIPTION
**Pull request by diml**
Following [this discussion](https://discuss.ocaml.org/t/its-hard-to-find-docs-and-examples-of-async-lib/2190/12) this PR changes the default build profile to be `dev`. The rationale being that most of the time we want to compile the things we work on in development mode.

The `-p` flags which must be used in opam file still implies `--profile release`.

Since this is technically a breaking changes, this behavior is only enabled for the `dune` binary. To do so this PR adds a module `Which_program` indicating which one of `dune` or `jbuilder` the current running program is. It uses the workaround for library variants described [here](https://github.com/diml/variants-workaround).

/cc @yminsky and @jberdine who tested the build profile feature

### CODE REVIEW

File: doc/usage.rst [hunk 1]
```diff
@@ -364,7 +364,7 @@ write a ``(profile ...)`` stanza. For instance:
 
 .. code:: scheme
 
-    (profile dev)
+    (profile performances)
```
Comment (emillon) [inline doc/usage.rst]: In that case, I believe that it is an is uncountable noun, and so we should use `performance`. Or maybe we can use a predefined profile name, like `(profile release)`?

File: test/blackbox-tests/test-cases/github660/run.t [hunk 2]
```diff
@@ -4,21 +4,26 @@ When there are explicit interfaces, modules must be rebuilt.
...
           main alias runtest
   hello
-  $ echo 'let x = 1' >> explicit-interfaces/lib_sub.ml
+  $ echo 'let _x = 1' >> explicit-interfaces/lib_sub.ml
   $ dune runtest --root explicit-interfaces
   Entering directory 'explicit-interfaces'
           main alias runtest
   hello
 
+CR-someday diml: this is weird that we need to do that:
```
Comment (emillon) [inline test/blackbox-tests/test-cases/github660/run.t]: That seems weird, what happens if this is not added? If that's not fixable immediately, I think we should replace this CR- comment by a link to a bug report.

Comment (diml) [reply test/blackbox-tests/test-cases/github660/run.t]: I added a test in https://github.com/ocaml/dune/pull/923

Comment (diml) [reply test/blackbox-tests/test-cases/github660/run.t]: For this PR I just changed the test to create the modified files in run.t

Comment (emillon) [reply test/blackbox-tests/test-cases/github660/run.t]: Yes, that seems better that way indeed!

**Pull request merged**